### ISSUE TITLE
Support for OAuth 2.0?

### ISSUE THREAD

**Issue by uber1geek** [freakboy3742/pyxero#276]
Is PyXero going to support OAuth 2.0? Xero has deprecated OAuth 1.0 in favor of the newer spec and all not allowing to create new private applications. There is no official Python SDK available as of today.
Any existing workarounds while that is official?

**Comment by TomskDiver** [comment]
@freakboy3742 may be see PR #275?

**Comment by sdolemelipone** [comment]
Or PR #278 for a version which is similar to existing Public/Partner credentials.

**Comment by uber1geek** [comment]
What about private credentials?

**Comment by sdolemelipone** [comment]
According [to xero](https://developer.xero.com/documentation/oauth2/overview) they "will not be providing a migration path for public and private apps". So it's OAuth2 only for new apps. You only need to go through the authentication process once then keep refreshing the token in order to get the same access as could be obtained with Private Credentials.

**Comment by dobesv** [comment]
BTW if you only need to generate an access token and refresh token for server-to-server access, you can use the Google OAuth2 playground:

https://developers.google.com/oauthplayground/#step3&scopes=accounting.transactions%20offline_access&url=https%3A%2F%2F&content_type=application%2Fjson&http_method=GET&useDefaultOauthCred=checked&oauthEndpointSelect=Custom&oauthAuthEndpointValue=https%3A%2F%2Flogin.xero.com%2Fidentity%2Fconnect%2Fauthorize&oauthTokenEndpointValue=https%3A%2F%2Fidentity.xero.com%2Fconnect%2Ftoken&includeCredentials=unchecked&accessTokenType=bearer&autoRefreshToken=unchecked&accessType=offline&prompt=consent&response_type=code&wrapLines=on

**Comment by dobesv** [comment]
To install the version in #278 without waiting for it to be merged and published you can install directly from git:

    pip install --upgrade 'git+https://github.com/sdolemelipone/pyxero.git@oauth2#egg=pyxero'

**Comment by Plantain** [comment]
@dobesv how did you use the access and refresh token from the Google Playground with #278? I can't see how you plug in those details without a full implementation as per the README.

**Comment by dobesv** [comment]
@Plaintain 

1. Click the link above
2. Fill in the OAuth Client ID and OAuth Client Secret
3. Put the scopes you want (e.g. `accounting.transactions offline_access`) in Step 1 on the left
4. Click "Authorize APIs" to initiate the OAuth flow
5. When you return to the playground, it shows the oauth code
6. Click to exchange the code for the refresh token and access token.
7. Copy these to a safe place and close that browser so it doesn't refresh the token and invalidate your refresh token.

Hopefully that makes sense, otherwise let me know what step you get stuck on.

**Comment by dobesv** [comment]
Oh wait, I misread your question.

When you construct the `OAuth2Credentials` object you can pass in a `token` parameter.  Pass in an object with the fields from the response to "exchange code for tokens" step, plus a calculated `expires_at` value.

```
token = dict(
  refresh_token='...',
  access_token='...',
  expires_in=1800,
  expires_at=time.time() + 1800
)
new_creds = OAuth2Credentials(client_id, client_secret, token=token)
```

**Comment by Plantain** [comment]
Thanks for replying, I did get that far - it's what to do with those tokens
next I'm stuck on, I don't see where I can use them without doing the whole
flow inside pyxero.

**Comment by Plantain** [comment]
Ok, that looks like the piece I was missing. Thankyou.

**Comment by Plantain** [comment]
I've followed all these steps, but still am still not able to use this patch with tenant errors:
>>> xero.contacts.all()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/plantain/xero/venv/lib/python3.8/site-packages/xero/basemanager.py", line 183, in wrapper
    raise XeroTenantIdNotSet
xero.exceptions.XeroTenantIdNotSet

Trying to specify the tenant doesn't seem to work either:
>>> creds.set_default_tenant()
Traceback (most recent call last):
  File "/home/plantain/xero/venv/lib/python3.8/site-packages/xero/auth.py", line 633, in set_default_tenant
    self.tenant_id = self.get_tenants()[0]['tenantId']
  File "/home/plantain/xero/venv/lib/python3.8/site-packages/xero/auth.py", line 626, in get_tenants
    self._handle_error_response(response)
  File "/home/plantain/xero/venv/lib/python3.8/site-packages/xero/auth.py", line 644, in _handle_error_response
    raise XeroUnauthorized(response)
KeyError: 'oauth_problem'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/plantain/xero/venv/lib/python3.8/site-packages/xero/auth.py", line 635, in set_default_tenant
    raise XeroException(None, "This xero client is not authorised for "
xero.exceptions.XeroException: This xero client is not authorised for any organisations.
>>> creds.get_tenants()

**Comment by Plantain** [comment]
I think the tenant problems are a red herring, I dumped out the actual error:
{"title":"Unauthorized","status":401,"detail":"TokenInvalidSignature","instance":"xxx"}

**Comment by dobesv** [comment]
Just keep in mind Xero has a really annoying token system, when you refresh, your old refresh token is invalidated, so you have to keep your tokens where you can update them regularly, and you might need some kind of locking system so that you don't try to refresh the token from two places at once, since one will mess up the other one.

**Comment by lmonty** [comment]
@dobesv I'm trying to use the google oauth 2.0 playground to get the same result but every time I am getting a 500 error. Any ideas on what might be going wrong? I've just started with xero and looking to integrate to the rest of my systems but I'm finding the auth flow very tricky!

**Comment by dobesv** [comment]
No I can't really tell just from status 500, check everything.  Make sure you are using the right client key and secret and you added the playground as an authorized redirect URL

**Comment by lmonty** [comment]
@dobesv Thanks, I hadn't set the playground as the authorised redirect URL. Now that I have my access_token, how best do I authenticate with the xero api?

**Comment by dobesv** [comment]
@lmonty Be sure to expand all the comments in this thread, I gave some helpful information to @Plantain above, hopefully that'll be enough to get you going as well.

**Comment by paladin0** [comment]
any chance this is being merged soon?

**Comment by freakboy3742** [comment]
This has now been merged, and released as part of v0.9.2.
